apps/tx-builder/src/components/Builder.tsx

```diff
@@ -0,0 +1,319 @@
+
+  const [toInput, setToInput] = useState('');
+  const [valueInput, setValueInput] = useState('');
+
+  const [transactions, setTransactions] = useState<ProposedTransaction[]>([]);
+  const [reviewing, setReviewing] = useState(false);
+  const [selectedMethodIndex, setSelectedMethodIndex] = useState(0);
+  const [showExamples, setShowExamples] = useState(false);
+  const [addTxError, setAddTxError] = useState<string | undefined>();
+  const [inputCache, setInputCache] = useState<string[]>([]);
+  const [isValueInputVisible, setIsValueInputVisible] = useState(false);
+
+  const handleMethod = useCallback(
+    async (methodIndex: number) => {
+      if (!contract || contract.methods.length <= methodIndex) return;
+      setSelectedMethodIndex(methodIndex);
+    },
+    [contract],
+  );
+
+  const handleInput = useCallback(
+    async (inputIndex: number, input: string) => {
+      inputCache[inputIndex] = input;
+      setInputCache(inputCache.slice());
+    },
+    [inputCache],
+  );
+
+  const getInputHelper = (input: any) => {
+    // This code renders a helper for the input text.
+    if (input.type.startsWith('tuple')) {
+      return `tuple(${input.components.map((c: any) => c.internalType).toString()})${
+        input.type.endsWith('[]') ? '[]' : ''
+      }`;
+    } else {
+      return input.type;
+    }
+  };
+
+  const getContractMethod = useCallback(() => contract?.methods[selectedMethodIndex], [contract, selectedMethodIndex]);
+
+  const handleSubmit = () => {
+    sendTransactions();
+    setTransactions([]);
+    setReviewing(false);
+  };
+
+  const handleDismiss = () => {
+    setReviewing(false);
+  };
+
+  const addTransaction = async () => {
+    let description = '';
+    let data = '';
+    const web3 = services.web3;
+
+    if (!web3) {
+      return;
+    }
+
+    if (contract && contract.methods.length > selectedMethodIndex) {
+      const method = contract.methods[selectedMethodIndex];
+      const cleanInputs = [];
+      description = method.name + ' (';
+      for (let i = 0; i < method.inputs.length; i++) {
+        const cleanValue = inputCache[i] || '';
+        cleanInputs[i] = cleanValue.charAt(0) === '[' ? JSON.parse(cleanValue.replace(/"/g, '"')) : cleanValue;
+        if (i > 0) {
+          description += ', ';
+        }
+        const input = method.inputs[i];
+        description += (input.name || input.type) + ': ' + cleanValue;
+      }
+      description += ')';
+
+      try {
+        if (!['receive', 'fallback'].includes(method.name)) {
+          data = web3.eth.abi.encodeFunctionCall(method as AbiItem, cleanInputs);
+        }
+      } catch (error) {
+        setAddTxError(error.message);
+        return;
+      }
+    }
+
+    try {
+      const cleanTo = web3.utils.toChecksumAddress(toInput);
+      const cleanValue = valueInput.length > 0 ? web3.utils.toWei(valueInput) : 0;
+      if (data.length === 0) {
+        data = '0x';
+      }
+      if (description.length === 0) {
```

Comment:
Shouldn't we rather check if data is 0?

Reply:
I think it will give the same result, but checking for `data` it's more clear.